src/parthenon_manager.cpp

```diff
@@ -209,69 +224,90 @@ void ParthenonManager::RestartPackages(Mesh &rm, RestartReader &resfile) {
-    const size_t v4 = v->GetDim(4);
-    const std::string vName = v->label();
-
-    if (Globals::my_rank == 0) std::cout << "Var:" << vName << ":" << v4 << std::endl;
-    // Read relevant data from the hdf file
-    int stat = resfile.ReadBlocks(vName.c_str(), myBlocks, tmp, bsize, v4);
-    if (stat < 0) {
-      std::cout << " WARNING: Variable " << v->label() << " Not found in restart file";
+
+  // make sure we have all sparse variables that are in the restart file
+  PARTHENON_REQUIRE_THROWS(
+      num_sparse == sparse_info.num_sparse,
+      "Mismatch between sparse fields in simulation and restart file");
+
+  std::vector<Real> tmp(static_cast<size_t>(nb) * nCells * max_vlen);
+  for (auto &v_info : indep_restart_vars) {
+    const auto vlen = v_info->NumComponents();
+    const auto &label = v_info->label();
+
+    if (Globals::my_rank == 0) std::cout << "Var:" << label << ":" << vlen << std::endl;
+    // Read relevant data from the hdf file, this works for dense and sparse variables
+    try {
+      resfile.ReadBlocks(label, myBlocks, tmp, bsize, vlen);
+    } catch (std::exception &ex) {
+      std::cout << "[" << Globals::my_rank << "] WARNING: Failed to read variable "
+                << label << " from restart file:" << std::endl
+                << ex.what() << std::endl;
       continue;
     }
 
     size_t index = 0;
     for (auto &pmb : rm.block_list) {
-      bool found = false;
-      const auto this_indep_restart_vars =
-          pmb->meshblock_data.Get()
-              ->GetVariablesByFlag(
-                  {parthenon::Metadata::Independent, parthenon::Metadata::Restart}, false)
-              .vars();
-      for (auto &v : this_indep_restart_vars) {
-        if (vName.compare(v->label()) == 0) {
-          auto v_h = v->data.GetHostMirror();
-
-          // Note index l transposed to interior
-          for (int k = out_kb.s; k <= out_kb.e; ++k) {
-            for (int j = out_jb.s; j <= out_jb.e; ++j) {
-              for (int i = out_ib.s; i <= out_ib.e; ++i) {
-                for (int l = 0; l < v_h.GetDim(4); ++l) {
-                  v_h(l, k, j, i) = tmp[index++];
-                }
-              }
+      if (v_info->IsSparse()) {
+        // check if the sparse variable is allocated on this block
+        if (sparse_info.IsAllocated(pmb->gid, sparse_idxs.at(label))) {
+          pmb->AllocateSparse(label);
+        } else {
+          // nothing to read for this block, advance reading index
+          index += nCells * vlen;
+          continue;
+        }
+      }
+
+      auto v = pmb->meshblock_data.Get()->GetCellVarPtr(label);
+      auto v_h = v->data.GetHostMirror();
+
+      // Note index l transposed to interior
+      for (int k = out_kb.s; k <= out_kb.e; ++k) {
+        for (int j = out_jb.s; j <= out_jb.e; ++j) {
+          for (int i = out_ib.s; i <= out_ib.e; ++i) {
+            for (int l = 0; l < vlen; ++l) {
+              v_h(l, k, j, i) = tmp[index++];
```

Comment:
I wasn't aware that we're making this index juggle.
Did someone test if there's an advantage for this?
IIRC this came up in the call with the HDF5 people with respect to writing larger chunks (or so), correct?